accepted/2021/experimental-package-shading/experimental-package-shading.md

```diff
@@ -0,0 +1,339 @@
+
+The preferred long-term solution is consumer-side shading, where shading is performed holistically for an app, where it shade packages only when necessary, and deduplicate compatible and identical shaded assemblies.
+
+Producer-side shading uses same building blocks as consumer-side shading, but is substantially smaller in scope. It also has the advantage that consumers of a producer-side shading package do not need tooling that understands shading.
+
+Developing and releasing producer-side shading as an experimental feature allows us to build out and test some of the underlying mechanisms for consumer-side shading, while making shading functionality available in the near term to package authors who need it to solve specific problems they are encountering.
+
+## Background
+
+### Dependency unification
+
+To understand why shading is needed, we must understand dependency unification and the problems it can cause.
+
+NuGet only allows a single version of each package to be resolved in the package graph. When a project has multiple references to the same package, NuGet must *unify* them to a single version. This frequently happens with transitive references to common packages such as `Newtonsoft.Json`.
+
+To demonstrate this, consider an example where a project references package `Foo v2.0` and package `Bar`, and `Bar` depends on `Foo v1.0`:
+
+```mermaid
+flowchart LR
+P[Project]
+P --> B[Bar]
+P --> |2.0| C2[Foo v2.0]
+B --> |1.0| C1[Foo v1.0]
+```
+
+The project *transitively* depends on `Foo v1.0` and directly depends on `Foo v2.0`. A project cannot depend on multiple versions of the same package, because this would cause conflicts and ambiguity. NuGet must *unify* the two differently versioned Foo dependencies to a single version in the context of that project.
+
+> **NOTE** Unification is independent for each project. All the direct and transitive dependencies for each project must be unified in the context of that project, but the result of unification is specific to that set of dependencies. Another project in the same solution with different dependencies and/or dependency versions may unify to different versions of the same dependencies.
+
+NuGet performs unification based on the dependency versions defined in the referencing packages or projects. Dependency versions may be exact, an explicit range, or a simple version that implicitly means "equal to or greater than". In the above example, the `v1.0` dependency means `>= v1.0`, so it is compatible with `v2.0`, and NuGet can unify them:
+
```

Comment:
I feel like we need a whole section talking about the difference between package versions and assembly versions.  Some packages don't rev their assembly versions at all and its worth noting that the file system only supports a single file in each directory with a unique name which reflects how NuGet unifies the dependency graph.  I don't want people to think that its bad design, its just how it is.